pest/src/position.rs

```diff
@@ -17,16 +17,34 @@ use std::str;
 use span;
 
 /// A cursor position in a `&str` which provides useful methods to manually parse that string.
+#[derive(Clone)]
 pub struct Position<'i> {
-    input: &'i [u8],
+    input: &'i str,
     pos: usize
 }
 
-pub unsafe fn new(input: &[u8], pos: usize) -> Position {
-    Position { input, pos }
-}
-
 impl<'i> Position<'i> {
+
+    /// Attempts to create a new `Position` at the given position. If the specified position is
+    /// an invalid index, or the specified position is not a valid UTF8 boundary, then None is
+    /// returned.
+    ///
+    /// # Examples
+    /// ```
+    /// # use pest::Position;
+    /// let cheart = '💖';
+    /// let heart = "💖";
+    /// assert_eq!(Position::new(heart, 1), None);
+    /// assert_ne!(Position::new(heart, cheart.len_utf8()), None);
+    /// ```
+    pub fn new(input: &str, pos: usize) -> Option<Position> {
+        if input.len() < pos || input.get(pos..).is_none() {
```

Comment:
This is now redundant. You can just do

```rust
input.get(pos..).map(|input| Position { input, pos })
```

because if `pos > input.len()`, `input.get(pos..)` is going to return `None`.

Reply:
Alright I removed the redundant code. Still experiencing some bug(s?) that is causing the calculator tests to fail, even though unit tests pass.